File: leetcode.js/三数之和.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function(nums) {
    const result = [];
    const n = nums.length;
    // 首先对数组进行排序
    nums.sort((a, b) => a - b)
    for (let i = 0; i < n - 2; i++) {
        // 如果当前数字大于0，则三数之和一定大于0，所以结束循环
        if (nums[i] > 0) break;
        // 去重
        if (i > 0 && nums[i] === nums[i - 1]) continue;

        let left = i + 1;
        let right = n - 1;
        while (left < right) {
            const sum = nums[i] + nums[left] + nums[right];
            if (sum === 0) {
                result.push([nums[i], nums[left], nums[right]]);
                // 去重
                while (left < right && nums[left] === nums[left + 1]) left++;
                while (left < right && nums[right] === nums[right - 1]) right--;
                left++;
                right--;
            } else if (sum < 0) {
                left++;
            } else {
                right--;
            }
        }
    }
    return result;
};
// 以下是使用 threeSum 函数的示例
const nums = [-1, 0, 1, 2, -1, -4];
const result = threeSum(nums);
console.log(result); // 输出: [ [ -1, -1, 2 ], [ -1, 0, 1 ] ]

File: leetcode.py/三数之和.py
class Solution(object):
    def threeSum(self, nums): 

        nums.sort()
        result = []
        length = len(nums)
        for i in range(length - 2):
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            left, right = i + 1, length - 1
            while left < right:
                total = nums[i] + nums[left] + nums[right]
                if total < 0:
                    left += 1
                elif total > 0:
                    right -= 1
                else:
                    result.append([nums[i], nums[left], nums[right]])
                    while left < right and nums[left] == nums[left + 1]:
                        left += 1
                    while left < right and nums[right] == nums[right - 1]:
                        right -= 1
                    left += 1
                    right -= 1
        return result
# 以下是使用 Solution 类的 threeSum 方法的示例
if __name__ == "__main__":
    solution = Solution()
    nums = [-1, 0, 1, 2, -1, -4]
    print("输入的数组为:", nums)
    print("三数之和为 0 的所有不重复三元组为:", solution.threeSum(nums))
